drivers/epson/projector/esc_vp21.cr

```diff
@@ -0,0 +1,274 @@
+require "digest/md5"
+require "placeos-driver/interface/muteable"
+require "placeos-driver/interface/powerable"
+require "placeos-driver/interface/switchable"
+
+module Epson; end
+
+module Epson::Projector; end
+
+# Documentation: documentation link
+
+class Epson::Projector::EscVp21 < PlaceOS::Driver
+  include Interface::Powerable
+  include Interface::Muteable
+
+  enum Inputs
+    HDMI
+    HDBaseT
+  end
+
+  # Discovery Information
+  tcp_port 1024
+  descriptive_name "Epson Projectors"
+  generic_name :Display
+
+  def on_load
+    transport.tokenizer = Tokenizer.new("\r\n")
+
+    self[:volume_min] = 0
+    self[:volume_max] = 255
+
+    self[:power] = 0
+    self[:stable_state] = true
+
+    self[:type] = :projector
+  end
+
+  def on_update
+  end
+
+  def connected
+    # Have to init comms
+    send("ESC/VP.net\x10\x03\x00\x00\x00\x00")
+    do_poll
+    schedule.every(52.seconds) { do_poll }
+  end
+
+  def disconnected
+    self[:power] = false
+    schedule.clear
+
+    @channel.close unless @channel.closed?
+  end
+
+  # used to coordinate the projector password hash
+  @channel : Channel(String) = Channel(String).new
+
+  #
+  # Power commands
+  #
+  def power(state : Bool)
+    self[:stable_state] = false
+    if state
+      self[:power_target] = true
+      do_send(:PWR, true, timeout: 40000, name: :power)
+      logger.debug { "-- epson Proj, requested to power on" }
+      do_send(:PWR, name: :power_state)
+    else
+      self[:power_target] = false
+      do_send(:PWR, false, timeout: 10000, name: :power)
+      logger.debug { "-- epson Proj, requested to power off" }
+      do_send(:PWR, name: :power_state)
+    end
+  end
+
+  def power?(**options, &block)
+    options[:emit] = block unless block.nil?
+    options[:name] = :power_state
+    do_send(:PWR, **options)
+  end
+
+  #
+  # Input selection
+  #
+  INPUTS = {
+    Inputs::HDMI    => 0x30,
+    Inputs::HDBaseT => 0x80,
+  }
+  INPUT_LOOKUP = INPUTS.invert
+
+  def switch_to(input : Inputs)
+    do_send(:SOURCE, INPUT_LOOKUP[input], name: :inpt_source)
+    do_send(:SOURCE, name: :inpt_query)
+
+    logger.debug { "-- epson LCD, requested to switch to: #{input}" }
+    self[:input] = input # for a responsive UI
+    self[:MUTE] = false
+  end
+
+  #
+  # Volume commands are sent using the inpt command
+  #
+  def volume(vol : Int32, **options)
+    vol = 0 if vol < 0
+    vol = 255 if vol > 255
```

Comment:
Take a look at [clamp](https://crystal-lang.org/api/0.35.1/Comparable.html#clamp(min,max)-instance-method) for this use case.